### PR TITLE
[TDD] Refactoring ML Service API for increasing maintainability

### PR DESCRIPTION
**Pull request by again4you**
To increase maintainability, this PR does refactoring for applying the Strategy pattern. The development process is TDD and detailed works are as below.

* Add Error code for DB and filesystem
* Add various test cases for ML Service API
* Refactoring the service API for maintainability and readability

### UML
PR_UML

#### V2
* Rename the delete API to `ml_service_delete_pipeline()`.
* Apply `Singleton` pattern to MLServiceLevelDB class.
* Add `const` key word for unchaged parameters.
* Update the error message in detail.
* Add the clean-up logic when MLServiceLevelDB instance is finalized.
* Raise `ML_ERROR_INVALID_PARAMETER` error when deleting key is not exists.

#### V1
* Update the signature of user-level APIs
* Add the positive and negative cases for setting, getting, and deleting the pipeline description
* Newly add the error code for database and file system.

### CODE REVIEW

File: c/src/ml-api-service-db.cc [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+/**
+ * @brief Class for implementation of IMLServiceDB
+ */
+class MLServiceLevelDB : public IMLServiceDB
+{
+public:
+  MLServiceLevelDB (std::string path);
+  virtual ~ MLServiceLevelDB ();
+  virtual void connectDB () override;
+  virtual void setPipelineDescription (std::string name,
```
Comment (jihochu) [inline c/src/ml-api-service-db.cc:32]: `name` 과 `pipeline_description` 이 함수내에서 변경되지 않는 것 같은데, const 를 붙여서 명확히 해주는게 어떨까요?
```suggestion
  virtual void setPipelineDescription (const std::string name, const std::string pipeline_description)
```

File: c/src/ml-api-service-db.cc [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+  return;
+}
+
+/**
+ * @brief Get the pipeline description with the given name.
+ * @param[in] name The unique name to retrieve.
+ * @param[out] pipeline_description The pipeline corresponding with the given name.
+ */
+void
+MLServiceLevelDB::getPipelineDescription (std::string name,
```
Comment (jihochu) [inline c/src/ml-api-service-db.cc:134]: 여기도 name 은 const 가 좋을 것 같습니다.

File: tests/capi/unittest_capi_service.cc [hunk 3]
```diff
@@ -8,14 +8,167 @@
...
+TEST (nnstreamer_capi_service, set_pipeline_description_1_n)
+{
+  int ret = 0;
+  const gchar *key = "ServiceName";
+  gchar *pipeline, *test_model;
+  const gchar *root_path = g_getenv ("NNSTREAMER_SOURCE_ROOT_PATH");
+
+  /* supposed to run test in build directory */
+  if (root_path == NULL)
+    root_path = "..";
```
Comment (jihochu) [inline tests/capi/unittest_capi_service.cc:58]: 테스트를 준비하는 동일한 코드가 많아 보입니다.
TEST_F 를 사용하여 class 화 하면 동일한 코드 사용을 줄이고 가독성을 높일 수 있을 것으로 생각됩니다.
한 번 고려해 보시기 바랍니다.

File: c/src/ml-api-service-db.cc [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+}
+
+/**
+ * @brief Construct a new MLServiceLevelDB object
+ * @param path database path
+ */
+MLServiceLevelDB::MLServiceLevelDB (std::string path)
+:  path (path)
+{
+  db_obj = nullptr;
```
Comment (jaeyun-jung) [inline c/src/ml-api-service-db.cc:69]: db_roptions and db_woptions also should be initialized as nullptr here to avoid svace issue.

File: c/src/ml-api-service-db.cc [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+};
+
+static IMLServiceDB *instance = nullptr;
+
+/**
+ * @brief Get the instance object of MLServiceLevelDB
+ * @return IMLServiceDB* instance of MLServiceLevelDB
+ */
+static IMLServiceDB *
+get_instance ()
```
Comment (jihochu) [inline c/src/ml-api-service-db.cc:52]: instance 를 하나로 제약하시려는 의도인 것 같은데, `MLServiceLevelDB` class 를 singleton 으로 만드는게 좋지 않을까 합니다.
각각의 MLServiceDB 가 singletom 이어야 한다면, IMLServiceDB 에서 singletom interface 을 옮기는 것도 추천 드립니다.

File: c/src/ml-api-service-db.cc [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+
+  return instance;
+}
+
+/**
+ * @brief Construct a new MLServiceLevelDB object
+ * @param path database path
+ */
+MLServiceLevelDB::MLServiceLevelDB (std::string path)
+:  path (path)
```
Comment (jihochu) [inline c/src/ml-api-service-db.cc:67]: 아래와 같이 수정하는 것을 추천드립니다.

```suggestion
:  path (path), db_obj(nullptr)
{
}
```

File: c/src/ml-api-service-db.cc (hunk 5 above)
Comment (jaeyun-jung) [inline c/src/ml-api-service-db.cc:52]: Discussion: make instance as static in MLServiceLevelDB class
declare static instance; in class MLServiceLevelDB

MLServiceLevelDB * MLServiceLevelDB::instatnce = nullptr;

MLServiceLevelDB *
MLServiceLevelDB::getInstance ()
{
 ...
}

File: c/include/ml-api-service.h [hunk 7]
```diff
@@ -77,22 +77,44 @@ typedef struct {
...
+ * @since_tizen 7.0
+ * @param[in] name The unique name to delete.
+ * @return @c 0 on success. Otherwise a negative error value.
+ * @note If the name did not exist in the database, this function returns ML_ERROR_NONE without any errors.
+ * @retval #ML_ERROR_NONE Successful.
+ * @retval #ML_ERROR_NOT_SUPPORTED Not supported.
+ * @retval #ML_ERROR_INVALID_PARAMETER Fail. The parameter is invalid.
+ * @retval #ML_ERROR_IO_ERROR The operation of DB or filesystem is failed.
+ */
+int ml_service_del_pipeline (const char *name);
```
Comment (jaeyun-jung) [inline c/include/ml-api-service.h:117]: recommend to change ml_service_delete_pipeline()

File: c/src/ml-api-service-db.cc [hunk 8]
```diff
@@ -0,0 +1,268 @@
...
+{
+  char *err = nullptr;
+  leveldb_options_t *db_options;
+
+  db_options = leveldb_options_create ();
+  leveldb_options_set_create_if_missing (db_options, 1);
+
+  db_obj = leveldb_open (db_options, path.c_str (), &err);
+  if (err != nullptr) {
+    _ml_loge ("Error! Failed to open Database : %s", err);
```
Comment (myungjoo) [inline c/src/ml-api-service-db.cc:93]: Error messages for API should be elaborated.

E.g.,
```
"Error! Failed to open database located at '%s': leveldb_open () has returned an error: %s", path.c_str(), err
```

File: c/src/ml-api-service-db.cc [hunk 9]
```diff
@@ -0,0 +1,268 @@
...
+void
+MLServiceLevelDB::setPipelineDescription (std::string name,
+    std::string pipeline_description)
+{
+  char *err = NULL;
+
+  leveldb_put (db_obj, db_woptions, name.c_str (), name.size (),
+      pipeline_description.c_str (), pipeline_description.size (), &err);
+  if (err != nullptr) {
+    _ml_loge ("Failed to put %s : %s", name.c_str (), err);
```
Comment (myungjoo) [inline c/src/ml-api-service-db.cc:120]:
```
"Failed to call leveldb_put () for the name, '%s' of the pipeline description (size: %d bytes / decription: '%.40s%s'). leveldb_put () has returned an error: %s", ... , (name.size() > 40) ? "...", ""
```

Try to give as much info as possible that may be used for debugging.

File: c/src/ml-api-service-db.cc [hunk 10]
```diff
@@ -0,0 +1,268 @@
...
+    std::string & pipeline_description)
+{
+  char *err = NULL;
+  char *value = NULL;
+  gsize read_len;
+
+  value = leveldb_get (db_obj, db_roptions, name.c_str (), name.size (),
+      &read_len, &err);
+  if (!value) {
+    _ml_loge ("Fail to find the key %s", name.c_str ());
```
Comment (myungjoo) [inline c/src/ml-api-service-db.cc:144]:
```
"Failed to find they key '%s' with leveldb_get for a pipeline description.", ...
```

File: c/src/ml-api-service-db.cc [hunk 11]
```diff
@@ -0,0 +1,268 @@
...
+    leveldb_free (err);
```
Comment (myungjoo) [inline c/src/ml-api-service-db.cc:145]: You didn't check for err here. You've checked ```!value``` only. If err is available, use it for error message, too!

Comment (myungjoo) [reply c/src/ml-api-service-db.cc:145]: Handle the case of (!value && err). (print all available info for app writers!)

File: c/src/ml-api-service-db.cc [hunk 12]
```diff
@@ -0,0 +1,268 @@
+/* SPDX-License-Identifier: Apache-2.0 */
+/**
+ * Copyright (c) 2022 Samsung Electronics Co., Ltd. All Rights Reserved.
+ *
+ * @file ml-api-service-db.c
+ * @date 07 Mar 2022
+ * @brief NNStreamer/Service C-API implementation
```
Comment (gichan-jang) [inline c/src/ml-api-service-db.cc:7]: Please check the descriptions of the doxgen tag.

File: c/src/ml-api-service-db.cc [hunk 13]
```diff
@@ -0,0 +1,268 @@
...
+  }
+
+  if (err != nullptr) {
+    _ml_loge ("Failed to get %s : %s", name.c_str (), err);
+    leveldb_free (err);
+    throw std::runtime_error ("Failed to getPipelineDescription()!");
+  }
+
+  pipeline_description = std::string (value, read_len);
+  return;
```
Comment (gichan-jang) [inline c/src/ml-api-service-db.cc:156]: Is there a reason to return? For the readability?

File: tests/capi/unittest_capi_service.cc [hunk 14]
```diff
@@ -49,6 +203,164 @@ TEST (nnstreamer_capi_service, basic_test_0_p)
...
+
+  ret = ml_service_set_pipeline(key, pipeline);
+  EXPECT_EQ (ML_ERROR_NONE, ret);
+
+  ret = ml_service_del_pipeline (key);
+  EXPECT_EQ (ML_ERROR_NONE, ret);
+
+  /* Test */
+  ret = ml_service_del_pipeline (key);
+  EXPECT_EQ (ML_ERROR_NONE, ret);
```
Comment (gichan-jang) [inline tests/capi/unittest_capi_service.cc:358]: Is it negative case?

File: c/include/ml-api-service.h (hunk 7 above)
Comment (again4you) [reply c/include/ml-api-service.h:117]: The name you recommended is more clear. I will fix it.

File: c/src/ml-api-service-db.cc (hunk 1 above)
Comment (again4you) [reply c/src/ml-api-service-db.cc:32]: 네. 그렇게 하는 것이 더 명확해 보이네요.

File: c/src/ml-api-service-db.cc (hunk 4 above)
Comment (again4you) [reply c/src/ml-api-service-db.cc:69]: @jihochu 님과 @jaeyun-jung 의 의견을 따라서 아래와 같이 수정하도록 하겠습니다. 
```cpp
MLServiceLevelDB::MLServiceLevelDB (std::string path)
:  path (path), db_obj(nullptr), db_roptions (nullptr), db_woptions (nullptr)
```

File: c/src/ml-api-service-db.cc (hunk 2 above)
Comment (again4you) [reply c/src/ml-api-service-db.cc:134]: Right! I will.

File: c/src/ml-api-service-db.cc (hunk 13 above)
Comment (again4you) [reply c/src/ml-api-service-db.cc:156]: In this case, just readability of code.

File: c/src/ml-api-service-db.cc (hunk 8 above)
Comment (again4you) [reply c/src/ml-api-service-db.cc:93]: Thank you for your advice. I will update the error message as detailed as I can.

File: c/src/ml-api-service-db.cc (hunk 9 above)
Comment (again4you) [reply c/src/ml-api-service-db.cc:120]: I will.

File: c/src/ml-api-service-db.cc (hunk 10 above)
Comment (again4you) [reply c/src/ml-api-service-db.cc:144]: I will.

File: c/src/ml-api-service-db.cc (hunk 11 above)
Comment (again4you) [reply c/src/ml-api-service-db.cc:145]: If the key is not found in the database, `leveldb_get()` returns `NULL` without an error message.
However, the err is set in case that a database operation error such as file corruption or permission.
Anyway, I will update the error message and clean up this code.

File: c/src/ml-api-service-db.cc (hunk 5 above)
Comment (again4you) [reply c/src/ml-api-service-db.cc:52]: I will apply `Singleton` pattern for this issue.
Thank you for your advice.

File: tests/capi/unittest_capi_service.cc (hunk 14 above)
Comment (again4you) [reply tests/capi/unittest_capi_service.cc:358]: You are right. It is not a negative test case. I will merge this into the positive one.

File: tests/capi/unittest_capi_service.cc (hunk 3 above)
Comment (again4you) [reply tests/capi/unittest_capi_service.cc:58]: ACR 일정으로 인하여 이번 PR에서는 이대로 유지하고,
말씀하신 Test Fixture 기반으로 Test Case를 작성하는 것은 다음 PR에서 수정하도록 하겠습니다.

File: c/src/ml-api-service-db.cc (hunk 6 above)
Comment (again4you) [reply c/src/ml-api-service-db.cc:67]: ACK!

File: c/src/ml-api-service-db.cc (hunk 5 above)
Comment (again4you) [reply c/src/ml-api-service-db.cc:52]: v2에서 Singleton Pattern을 적용하였습니다.

File: c/src/ml-api-service-db.cc (hunk 12 above)
Comment (again4you) [reply c/src/ml-api-service-db.cc]: Updated! Thanks.

**Comment by again4you** [comment]
#### V2
* Rename the delete API to `ml_service_delete_pipeline()`.
* Apply `Singleton` pattern to MLServiceLevelDB class.
* Add `const` key word for unchaged parameters.
* Update the error message in detail.
* Add the clean-up logic when MLServiceLevelDB instance is finalized.
* Raise `ML_ERROR_INVALID_PARAMETER` error when deleting key is not exists.

File: tests/capi/unittest_capi_service.cc (hunk 3 above)
Comment (jihochu) [reply tests/capi/unittest_capi_service.cc:58]: 네~

**Review by jihochu — APPROVED** [review]
패턴을 적용하여 코드가 훨씬 가독성이 높아진것 같습니다.
👍

**Review by myungjoo — APPROVED** [review]
LGTM!

**Review by gichan-jang — APPROVED** [review]
LGTM!

File: tests/capi/unittest_capi_service.cc [hunk 15]
```diff
@@ -59,7 +338,7 @@ main (int argc, char **argv)
 
   try {
     testing::InitGoogleTest (&argc, argv);
-  } catch (...) {
+  } catch ( ...) {
```
Comment (gichan-jang) [inline tests/capi/unittest_capi_service.cc:341]: unnecessary change?

**Review by jaeyun-jung — APPROVED** [review]
